Automattic/woocommerce-payments
# Cancel non-captured authorizations if order is cancelled.

Fixes #5605

#### Changes proposed in this Pull Request

Cancel non-captured authorizations if order is cancelled.

When WooPayments plugin is running in "Issue an authorization on checkout, and capture later" mode, it creates an authorization when order is placed. Such authorization can later be captured from order screen or transactions screen.

However, when order is cancelled in WooCommerce, such authorizations remained, and could lead to a confusion, e.g. being shown as "Uncaptured" in the transactions screen.

Since there is no money transferred from the shopper to the merchant at this moment, we can fully cancel such authorizations when order is cancelled and remove any potential confusion.

Captured authorizations are not subject of this change, since in this case it is up to the merchant to decide how those should be refunded.

In line with #5712 when merchant tries to cancel an order with uncaptured authorization, a modal will appear, asking for confirmation.

**Before: modal confirmation**

_This modal still remains in use for paid orders which can be refunded._

![Before: Modal confirmation](https://github.com/Automattic/woocommerce-payments/assets/683297/1a0f606f-17f3-4151-9f72-fe065d7e19c9)

**Before: cancelled order**

![Before: Cancelled order](https://github.com/Automattic/woocommerce-payments/assets/683297/ba498454-60d8-4e01-ada5-c2237d22709a)

**Before: transactions screen**

<img width="379" alt="Before: Transactions screen" src="https://github.com/Automattic/woocommerce-payments/assets/683297/b166f4f3-8e03-465a-8616-6ece028b1c2e">

**Before: transaction details**

![Transaction details](https://github.com/Automattic/woocommerce-payments/assets/683297/58333421-cded-4bd7-91b0-4ff29030ea43)

**After: modal confirmation**

_Missing border line is a compression artifact_ 🤷‍♂️ 

![After: Modal confirmation](https://github.com/Automattic/woocommerce-payments/assets/683297/72a8ec7c-248e-45eb-a850-ef27778ee9cd)

**After: cancelled order**

![After: cancelled order](https://github.com/Automattic/woocommerce-payments/assets/683297/8282bdac-9472-479b-b1a2-11dc6f089039)

**After: transactions screen**

<img width="384" alt="After: Transactions screen" src="https://github.com/Automattic/woocommerce-payments/assets/683297/d12ea1e3-3eed-4f0a-947c-6332d3ee2a82">


#### Testing instructions

 * Enable _"Issue an authorization on checkout, and capture later"_ in the _Payments_ -> _Settings_ -> _Transactions_ section.
 * Create an order and checkout.
   * Authorization should be indicated in order notes and appear on the _Payments_ -> _Transactions_ screen.
 * Cancel an order.
   * Order notes should indicate cancelled authorization.
   * Authorization shouldn't appear on the _Payments_ -> _Transactions_ screen.

 * Disable _"Issue an authorization on checkout, and capture later"_
 * Create an order and checkout.
   * Successful transaction should appear in the _Payments_ -> _Transactions_ screen.
 * Cancel an order
   * Successful transaction should remain in the _Payments_ -> _Transactions_ screen.

-------------------


**Post merge**
